File: frontend/lib/permissions.ts
export enum Permission {
  // User permissions
  VIEW_DASHBOARD = 'view_dashboard',
  VIEW_PRAYER_TIMES = 'view_prayer_times',
  VIEW_SETTINGS = 'view_settings',
  CHANGE_PASSWORD = 'change_password',
  
  // Admin permissions
  MANAGE_USERS = 'manage_users',
  MANAGE_MOSQUES = 'manage_mosques',
  MANAGE_SETTINGS = 'manage_settings',
  VIEW_ADMIN_PANEL = 'view_admin_panel',
  
  // System permissions
  ACCESS_API = 'access_api',
  VIEW_LOGS = 'view_logs',
  MANAGE_SYSTEM = 'manage_system'
}

export enum UserRole {
  GUEST = 'guest',
  USER = 'user',
  ADMIN = 'admin'
}

// Permission mappings for each role
export const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  [UserRole.GUEST]: [
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_PRAYER_TIMES
  ],
  [UserRole.USER]: [
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_PRAYER_TIMES,
    Permission.VIEW_SETTINGS,
    Permission.CHANGE_PASSWORD,
    Permission.ACCESS_API
  ],
  [UserRole.ADMIN]: [
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_PRAYER_TIMES,
    Permission.VIEW_SETTINGS,
    Permission.CHANGE_PASSWORD,
    Permission.ACCESS_API,
    Permission.MANAGE_USERS,
    Permission.MANAGE_MOSQUES,
    Permission.MANAGE_SETTINGS,
    Permission.VIEW_ADMIN_PANEL,
    Permission.VIEW_LOGS,
    Permission.MANAGE_SYSTEM
  ]
};

// Helper functions
export function hasPermission(userRole: UserRole, permission: Permission): boolean {
  return ROLE_PERMISSIONS[userRole]?.includes(permission) ?? false;
}

export function getUserPermissions(userRole: UserRole): Permission[] {
  return ROLE_PERMISSIONS[userRole] ?? [];
}

export function canAccessAdminPanel(userRole: UserRole): boolean {
  return hasPermission(userRole, Permission.VIEW_ADMIN_PANEL);
}

export function canManageUsers(userRole: UserRole): boolean {
  return hasPermission(userRole, Permission.MANAGE_USERS);
}

export function canManageMosques(userRole: UserRole): boolean {
  return hasPermission(userRole, Permission.MANAGE_MOSQUES);
} 

File: backend/internal/model/permissions.go
package model

// Permission represents a specific permission in the system
type Permission string

const (
	// User permissions
	ViewDashboard   Permission = "view_dashboard"
	ViewPrayerTimes Permission = "view_prayer_times"
	ViewSettings    Permission = "view_settings"
	ChangePassword  Permission = "change_password"

	// Admin permissions
	ManageUsers    Permission = "manage_users"
	ManageMosques  Permission = "manage_mosques"
	ManageSettings Permission = "manage_settings"
	ViewAdminPanel Permission = "view_admin_panel"

	// System permissions
	AccessAPI    Permission = "access_api"
	ViewLogs     Permission = "view_logs"
	ManageSystem Permission = "manage_system"
)

// UserRole represents the role of a user
type UserRole string

const (
	Guest       UserRole = "guest"
	RegularUser UserRole = "user"
	Admin       UserRole = "admin"
)

// RolePermissions maps each role to its permissions
var RolePermissions = map[UserRole][]Permission{
	Guest: {
		ViewDashboard,
		ViewPrayerTimes,
	},
	RegularUser: {
		ViewDashboard,
		ViewPrayerTimes,
		ViewSettings,
		ChangePassword,
		AccessAPI,
	},
	Admin: {
		ViewDashboard,
		ViewPrayerTimes,
		ViewSettings,
		ChangePassword,
		AccessAPI,
		ManageUsers,
		ManageMosques,
		ManageSettings,
		ViewAdminPanel,
		ViewLogs,
		ManageSystem,
	},
}

// HasPermission checks if a role has a specific permission
func HasPermission(role UserRole, permission Permission) bool {
	permissions, exists := RolePermissions[role]
	if !exists {
		return false
	}

	for _, p := range permissions {
		if p == permission {
			return true
		}
	}
	return false
}

// GetUserPermissions returns all permissions for a given role
func GetUserPermissions(role UserRole) []Permission {
	return RolePermissions[role]
}

// CanAccessAdminPanel checks if a role can access the admin panel
func CanAccessAdminPanel(role UserRole) bool {
	return HasPermission(role, ViewAdminPanel)
}

// CanManageUsers checks if a role can manage users
func CanManageUsers(role UserRole) bool {
	return HasPermission(role, ManageUsers)
}

// CanManageMosques checks if a role can manage mosques
func CanManageMosques(role UserRole) bool {
	return HasPermission(role, ManageMosques)
}

// CanViewSettings checks if a role can view settings
func CanViewSettings(role UserRole) bool {
	return HasPermission(role, ViewSettings)
}

// CanChangePassword checks if a role can change password
func CanChangePassword(role UserRole) bool {
	return HasPermission(role, ChangePassword)
}
